### ISSUE TITLE
Unexpected (non)termination of tasks created by an allocator.

### ISSUE THREAD

**Issue by Richard-Wai** [Ada-Rapporteur-Group/User-Community-Input#82]
9.3(1) of the RM says:
Each task (other than an environment task — see 10.2) depends on one or more masters (see 7.6.1), as follows:
- If the task is created by the evaluation of an allocator for a given named access type, it depends on each master that includes the elaboration of the declaration of the ultimate ancestor of the given access type.
...
9.3(6/1) then says:
Completion of a task (and the corresponding [task_body](http://www.ada-auth.org/standards/22aarm/html/AA-9-1.html#S0248)) can occur when the task is blocked at a [select_statement](http://www.ada-auth.org/standards/22aarm/html/AA-9-7.html#S0269) with an open [terminate_alternative](http://www.ada-auth.org/standards/22aarm/html/AA-9-7-1.html#S0275) (see [9.7.1](http://www.ada-auth.org/standards/22aarm/html/AA-9-7-1.html)); the open [terminate_alternative](http://www.ada-auth.org/standards/22aarm/html/AA-9-7-1.html#S0275) is selected if and only if the following conditions are satisfied: 

- {[AI95-00415-01](http://www.ada-auth.org/cgi-bin/cvsweb.cgi/AIs/AI-00415.TXT)} The task depends on some completed master; and
- Each task that depends on the master considered is either already terminated or similarly blocked at a [select_statement](http://www.ada-auth.org/standards/22aarm/html/AA-9-7.html#S0269) with an open [terminate_alternative](http://www.ada-auth.org/standards/22aarm/html/AA-9-7-1.html#S0275).

Consider a partition with a lone compilation unit thus:

```Ada
with Ada.Text_IO; use Ada.Text_IO;
with Ada.Unchecked_Deallocation;

procedure Task_Term is
   task type Temp_Task is
      entry Sync;
   end;

   task body Temp_Task is
   begin
      Put_Line ("Delay");
      delay 2.0;
      Put_Line ("Done");

      select
         accept Sync;
      or
         terminate;
      end select;
   end;

   type Task_Access is access Temp_Task;

   procedure Free is new Ada.Unchecked_Deallocation (Temp_Task, Task_Access);

   Some_Task: Task_Access := new Temp_Task;

begin
   Free (Some_Task);
   Put_Line ("Terminated");
end;
```

The actual output of this program, according to the language rules should be:
```
Delay
Terminated
(~2 second pause)
Done
```

And this is indeed what GNAT does.

However I would wager that this is not at all what the causal Ada programmer would expect. Which would be:

```
Delay
(~2 second pause)
Done
Terminated
```

I think it is fair to say that most average users would expect  that Free will cause the Temp_Task allocated to Some_Task to then terminate. The language rules make it clear, however, that the selection of the terminate alternative is not available, since the dependent master of the task (the elaboration Task_Access) is not "complete" until leaving Task_Term.

I recognize the provision for exactly this as well in 13.11.2(9/5): "There is one exception: if the object being freed contains tasks, it is unspecified whether the object is deallocated.", but looking at the associated AIs suggests that this exists for a totally different reason.

It seems to reason that if a task created by an allocator is activated and begins executing immediately, that a call to Unchecked_Deallocation should behave in a similar fashion - waiting for the completion of the deallocated task and any dependents, and potentially the selection of any terminate alternatives in that chain.

I can't quite understand why this should not be done. I can understand why tasks created by object declarations should ensure that no other task dependent on the same master will be able to rendezvous with a task that no longer exists, but in the case of a task created through the evaluation of an allocator, the only way for such a thing to happen is through the incorrect use of Unchecked_Deallocation, which is true for any allocated object, not just task objects. Imagine an access type designating a class-wide synchronized interface - why should it be more safe if implemented by a task type rather than a protected type?

I think it really ought to be that an invocation of Unchecked_Deallocation on an access type designating a task type, or an object containing tasks, should wait for completion of all dependent tasks, including the selection of a terminate alternative if available.

Any issues that might arise from that behavior, in my mind, should be included as part of Unchecked_Deallocation's Erroneous Execution section. Curiously there is an awkward proivision hinting at the existing surprising semantics in the descrition of bounded errors at 13.11.2(11), where it is a bounded error to free a discriminated, unterminated task because the discriminants might be deallocated before the task!

I don't see huge risk in such a change, as I'd imagine there would be more code expecting my proposed semantics than relying on the existing semantics.

**Comment by Richard-Wai** [comment]
Obviously making changes like this is pushing a boulder up a hill, but This just occurred to me: 

What if instead we introduced the notion of an "allocated master". That is a master that is created during, and associated with, the evaluation of an allocation, and becomes the master of all objects that arise through the evaluation of said allocator

Then as part of Unchecked_Deallocation, that master would complete, and that would entail all the usual things, followed by deallocation of the allocated object, and then the allocated master.

We could do through through an aspect, though I prefer a key word in this particular case:

```Ada
type Task_Access is synchronized access Task_Type;
```

**Comment by CKWG** [comment]
> I think it is fair to say that most average users would expect that Free will cause the Temp_Task allocated to Some_Task to then terminate. The language rules make it clear, however, that the selection of the terminate alternative is not available, since the dependent master of the task (the elaboration Task_Access) is not "complete" until leaving Task_Term.

I do not think that I'm an exceptional Ada user, but I wondered what should happen in your code. Deallocating a task while it is running is a develish idea! I would never ever do that. And I remember vaguely that the RM has some caveats with respect to deallocating tasks. Some of them Richard quoted.

**Comment by jprosen** [comment]
Imagine the deallocated task is on a select-with-terminate, but it has subtasks that are still running. What would be your "expected" or "natural" behaviour in that case? You are asking for a special case where the only master that prevents termination is the one that executes the Unchecked_Deallocation. This seems to me to be extremely difficult to formalize for a very small benefit.
And I am not convinced that what you suggest is the expected behaviour by casual users. Yes, you have to understand the rules of the language - this is true of any language.

**Comment by sttaft** [comment]
The whole notion of the terminate alternative was an attempt to allow tasks to act somewhat like passive servers, but I think it, along with the whole rendezvous concept, has over time emerged as an interesting but failed experiment.  There are still a few interesting situations where rendezvous, and even terminate alternatives, are useful, but in the vast majority of cases, protected objects and/or explicit "shutdown" commands seem more understandable and manageable.  Given that, it seems hard to justify making significant investments in the "selective accept with terminate alternative" feature.

**Comment by jprosen** [comment]
Sorry, but I have a totally different view. Tasks are active objects, and the rendezvous is an easily understandable, high level way to communicate. Too high level for some low level simple synchonization mechanisms, and PO were invented to fill that gap. But as soon as you move from bare metal to higher abstractions, tasks and rendezvous are more appropriate.

**Comment by sttaft** [comment]
> Sorry, but I have a totally different view. Tasks are active objects, and the rendezvous is an easily understandable, high level way to communicate. Too high level for some low level simple synchonization mechanisms, and PO were invented to fill that gap. But as soon as you move from bare metal to higher abstractions, tasks and rendezvous are more appropriate.

I certainly understand that view, but my sense is that Ada programmers have largely moved away from rendezvous as a way to synchronize tasks, perhaps because it results in too tight a coupling between tasks, compared to the use of protected objects.  It would be interesting to gather some statistics.

**Comment by Richard-Wai** [comment]
Of course I can only speak for myself but I use them fairly heavily. I find in a very multi-threaded application, tasks often need that explicit synchronization that is much more ergonomic than using a PO, which basically just adds extra steps.

Task rendezvous can be used in place of "condition variables" as are commonly used in C++ to do tag-team like synchronization. Of course POs can do this exactly with entries, but that is often just extra steps.

I also found myself deploying a pattern that I loosely call "service request", which allows a task to do dual-duty of dequeuing from a queue most of the time, but then handling special synchronized actions outside of the queue that cannot be done easily with normal sideband stuff. An example is a fan-out task that distributes an input to a collection of other queues. This pattern allows one to efficiently implement subscribe/unsubscribe operations, for example. This pattern would be very unwieldy to do in any other language.

**Comment by sttaft** [comment]
> Of course I can only speak for myself but I use them fairly heavily. I find in a very multi-threaded application, tasks often need that explicit synchronization that is much more ergonomic than using a PO, which basically just adds extra steps. ...

OK, that's a useful data point!
